Simple tuple construction and underlying field access

VSadov opened the pull request:
Some very simple scenarios now work e2e from parsing to codegen.
There are some "PROTOTYPE" comments, but I feel we need to sync up, since this change opens many test scenarios.

VSadov commented:
@dotnet/roslyn-compiler - please take a look

jcouv commented on src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:
```diff
@@ -639,6 +626,63 @@ private BoundExpression BindExpressionInternal(ExpressionSyntax node, Diagnostic
+            {
+                var argumentSyntax = arguments[i];
+                var name = argumentSyntax.NameColon?.Name.Identifier.ValueText;
+
```
Is this where we'll put the logic to prevent using "Item1", "Item2", ... as friendly names?

VSadov replied:
For the tuple types that result from a tuple creation expression - Yes. 
Note that generally it would be an error to provide names just for some fields. The code here still tries to handle partial named scenario for error tolerance. I am not sure yet if that scenario is going to be ultimately reachable, but I want to err on the safe side. If any of the names are missing (for whatever reason) they would become "ItemX" since that seems the only sensible way to handle such cases if they happen as a result of error recovery or something like that.

jcouv commented on src/Compilers/CSharp/Portable/Binder/Binder_Lookup.cs:
```diff
@@ -187,6 +187,10 @@ protected void LookupMembersInType(LookupResult result, TypeSymbol type, string
 
+                case TypeKind.Tuple:
+                    useSiteDiagnostics = LookupMembersInTuple(result, type, name, arity, basesBeingResolved, options, originalBinder, diagnose, useSiteDiagnostics);
+                    break;
```
Why not use `ref useSiteDiagnostics` like neighboring code? It would be more consistent.

VSadov replied:
I did this through refactoring and did not notice the change.
I am not yet sure if we need a separate LookupMembersInTuple. So far it ended up just deferring to LookupMembersInClass. 
I think we might do the first iteration (before going into bases) in the LookupMembersInTuple and then call into LookupMembersInClass for underlying type, since that would be the same. The only gain would be that the special case check for tuples could be done only in LookupMembersInTuple and not affect other types that are not tuples.
For now I did not want to complicate things though, so LookupMembersInTuple just calls LookupMembersInClass.

VSadov replied:
made it "ref"

jcouv replied:
Yes, I was not questioning having a `LookupMembersInTuple`, just the `ref` part. Thanks

jcouv commented on src/Compilers/CSharp/Portable/Symbols/NamedTypeSymbol.cs:
```diff
@@ -634,15 +634,34 @@ public override int GetHashCode()
         internal override bool Equals(TypeSymbol t2, bool ignoreCustomModifiersAndArraySizesAndLowerBounds = false, bool ignoreDynamic = false)
         {
-            if (ReferenceEquals(this, t2)) return true;
+            if ((object)t2 ==  this) return true;
```
Extra space after `==`.

VSadov replied:
done

jcouv commented on src/Compilers/CSharp/Portable/Symbols/NamedTypeSymbol.cs:
```diff
@@ -634,15 +634,34 @@ public override int GetHashCode()
+
+                tupleType = this as TupleTypeSymbol;
+                if (tupleType != null)
+                {
```
Shouldn't the case where `this` is a `TupleTypeSymbol` be handled in an overridden `Equals` method on `TupleTypeSymbol` itself?

VSadov replied:
Yes, I think both checks above can move up to the TupleTypeSymbol override.

jcouv replied:
I'm not sure about the first of the two checks. That one may be in the right place. I'll stop by to discuss.

VSadov replied:
Right. The first is in the right place. We can enter with this not being tuple, but with t2 being a tuple that is equal.

jcouv commented on src/Compilers/CSharp/Portable/Symbols/TupleTypeSymbol.cs:
```diff
@@ -145,14 +143,14 @@ internal sealed override ObsoleteAttributeData ObsoleteAttributeData
         {
-            // TODO: members
-            return ImmutableArray<Symbol>.Empty;
+            //TODO: PERF do we need to have a dictionary here?
```
No TODOs. Mark as PROTOTYPE.

VSadov replied:
fixed

jcouv commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTupleTest.cs:
```diff
@@ -0,0 +1,167 @@
+    public class CodeGenTupleTests : CSharpTestBase
+    {
+
+
```
Extra empty lines

jcouv commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTupleTest.cs:
```diff
@@ -0,0 +1,167 @@
+{
+  // Code size       54 (0x36)
+  .maxstack  5
+  .locals init (System.Runtime.CompilerServices.ValueTuple<int, string> V_0, //x
```
Where did this `ValueTuple<int, string>` come from? There are only integers, but no strings, in the tuples above.

jcouv replied:
Nevermind, I misparsed the statement on line 85 ;-)

jcouv commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTupleTest.cs:
```diff
@@ -0,0 +1,167 @@
+  IL_0004:  ldc.i4.3
+  IL_0005:  ldc.i4.4
+  IL_0006:  newobj     ""System.Runtime.CompilerServices.ValueTuple<int, int>..ctor(int, int)""
+  IL_000b:  newobj     ""System.Runtime.CompilerServices.ValueTuple<int, >..ctor(int, )""
```
This `<int, >` seems strange. Is it correct?
I would think this should be `ValueTuple<int, ValueTuple<int, int>>`.

VSadov replied:
Something seems off with visualization of nested tuples (the type in actual IL seems correct). 
Will look at this in a later change.

jcouv commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTupleTest.cs:
```diff
@@ -0,0 +1,167 @@
+");
+        }
+
+
```
Extra empty line.

jcouv commented:
LGTM

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:
```diff
@@ -639,6 +626,63 @@ private BoundExpression BindExpressionInternal(ExpressionSyntax node, Diagnostic
+            if (arguments.Count < 2)
+            {
+                // this should be a parse error already.
+                return BadExpression(node);
```
We should bind arguments anyway and capture result a s children of the BadExpression, this is needed for SemanticModel, IOperation, etc.

VSadov replied:
Right. I am not sure this code will ultimately be reachable, but in case it is (for error recovery, maybe) the arguments must be bound.

VSadov replied:
done

gafter replied:
We might consider defining the syntax `(a: 1)` to create a oneple.

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:
```diff
@@ -639,6 +626,63 @@ private BoundExpression BindExpressionInternal(ExpressionSyntax node, Diagnostic
+                    }
+                }
+
+                var boundArgument = BindExpression(argumentSyntax.Expression, diagnostics);
```
We should enforce that this is a value.

VSadov replied:
fixed

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:
```diff
@@ -639,6 +626,63 @@ private BoundExpression BindExpressionInternal(ExpressionSyntax node, Diagnostic
+
+                var boundArgument = BindExpression(argumentSyntax.Expression, diagnostics);
+                boundArguments.Add(boundArgument);
+                elementTypes.Add(boundArgument.Type);
```
Do we need to specially handle type-less expressions, like ```null```, lambda, etc.

VSadov replied:
Good question. Null may be treated as object and we could use some kind of target typing. I'd make this simpler and just require these to be values. At least for now. User can add casts.

AlekseyTs replied:
Please add a test with null.

gafter replied:
I believe the spec will say that a tuple literal expression does not have a type, but is convertible to one of the System.Tuple or System.ValueTuple types through a *tuple conversion*. In other words it is a kind of target typing. I think that is the right approach to prototype it too, as we'll need experience with it. The individual members don't have to have a type either (e.g. `null` is OK), but the whole tuple literal's convertibility to a given target type will depend on the convertibility of each element to the corresponding tuple type's element.

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_Lookup.cs:
```diff
@@ -715,6 +725,15 @@ internal virtual bool SupportsExtensionMethods
+                // PROTOTYPE consider taking this out of the loop. tuples are always at the end of hierarchy
+                // PROTOTYPE match this in other lookups
+                var tuple = currentType as TupleTypeSymbol;
+                if (tuple != null)
```
(object)?

VSadov replied:
done

jcouv commented on src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:
```diff
@@ -639,6 +626,63 @@ private BoundExpression BindExpressionInternal(ExpressionSyntax node, Diagnostic
+                node,
+                this,
+                diagnostics);
+
```
Didn't you run into a problem with the `TupleTypeSymbol` constructor here? I think the code currently expects `elementNames` to be null or populated, but not `Empty`.

VSadov replied:
it looks like symbol construction expects empty, but that can be changed to default if needed

VSadov replied:
Default is probably a tiny bit cheaper to use here

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_Lookup.cs:
```diff
@@ -715,6 +725,15 @@ internal virtual bool SupportsExtensionMethods
+                var tuple = currentType as TupleTypeSymbol;
+                if (tuple != null)
+                {
+                    currentType = tuple.UnderlyingTupleType;
```
Similar change should probably done in AddMemberLookupSymbolsInfoInType

VSadov replied:
I've added a "PROTOTYPE" since this is not in scope of this particular change. Will definitely need that when implementing symbols info.

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Semantics/AccessCheck.cs:
```diff
@@ -221,6 +224,11 @@ private static bool IsNamedTypeAccessible(NamedTypeSymbol type, Symbol within, r
 
+            // for tuples consider the underlying type as a true container for visibility puroposes.
+            containingType = 
+                (containingType as TupleTypeSymbol)?.UnderlyingTupleType ?? 
```
We prefer kind checks to type checks.

VSadov replied:
changed

AlekseyTs commented on src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml:
```diff
@@ -1140,6 +1140,12 @@
+    <Node Name="BoundTupleCreationExpression" Base="BoundExpression">
+    <!-- Non-null type is required for this node kind -->
+    <Field Name="Type" Type="TypeSymbol" Override="true" Null="disallow"/>
+    <Field Name="Constructor" Type="MethodSymbol" Null="disallow"/>
```
Consider adding a comment about what constructor is this.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/NamedTypeSymbol.cs:
```diff
@@ -634,15 +634,34 @@ public override int GetHashCode()
+                //PROTOTYPE: rename ignoreDynamic or introduce another "ignoreTuple" flag
+                // if ignoring dynamic, compare underlying tuple types
+                var tupleType = t2 as TupleTypeSymbol;
+                if (tupleType != null)
```
We prefer Kind checks to type checks.

AlekseyTs replied:
I thought TupleTypeSymbol has its own kind. Cast to object then?

VSadov replied:
I thought we would need to cast anyways , so why not use trycast
Checking the kind and then casting would also be ok.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/NamedTypeSymbol.cs:
```diff
@@ -634,15 +634,34 @@ public override int GetHashCode()
+                }
+
+                tupleType = this as TupleTypeSymbol;
+                if (tupleType != null)
```
(object)?

VSadov replied:
no longer applicable

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/TupleTypeSymbol.cs:
```diff
@@ -303,6 +308,158 @@ public override bool IsSealed
+        {
+            get
+            {
+                return _underlyingType.ConstructedFrom;
```
This is wrong.

VSadov replied:
fixed

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/TupleTypeSymbol.cs:
```diff
@@ -303,6 +308,158 @@ public override bool IsSealed
+        {
+            get
+            {
+                return _underlyingType.MangleName;
```
How can you mangle an empty name.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/TupleTypeSymbol.cs:
```diff
@@ -303,6 +308,158 @@ public override bool IsSealed
+        {
+            get
+            {
+                return _underlyingType.Arity;
```
I would expect this to be 0.

VSadov replied:
Did not give a lot of thought to generics yet. I agree, modeling tuple as nongeneric type makes sense

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/TupleTypeSymbol.cs:
```diff
@@ -303,6 +308,158 @@ public override bool IsSealed
+        {
+            get
+            {
+                return _underlyingType.TypeParameters;
```
I would expect this to be an empty array.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/TupleTypeSymbol.cs:
```diff
@@ -303,6 +308,158 @@ public override bool IsSealed
+        {
+            get
+            {
+                return _underlyingType.HasSpecialName;
```
False?

VSadov replied:
made it false.
I think many emit related properties are not supposed to be used since tuple types are not going to be emitted. It is hard to tell whether properties like IsSerializable should defer to underlying or not. It probably still should even though it may not be serializable regardless.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/TupleTypeSymbol.cs:
```diff
@@ -341,6 +498,51 @@ public override TResult Accept<TResult>(SymbolVisitor<TResult> visitor)
+
+        internal override ImmutableArray<NamedTypeSymbol> GetInterfacesToEmit()
+        {
+            throw new NotImplementedException();
```
All this throws should be tracked, either by a comment, or an issue.

VSadov replied:
I will provide implementations.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/TupleTypeSymbol.cs:
```diff
@@ -341,6 +498,51 @@ public override TResult Accept<TResult>(SymbolVisitor<TResult> visitor)
+
+        internal override NamedTypeSymbol GetDeclaredBaseType(ConsList<Symbol> basesBeingResolved)
+        {
+            throw new NotImplementedException();
```
We need to decide what would be considered a base type for this symbol.

VSadov replied:
it should be same as the base of the underlying type. We will not use base for the purpose of lookup, but may still need the base when analyzing casting.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/TupleTypeSymbol.cs:
```diff
@@ -89,16 +91,12 @@ internal sealed class TupleTypeSymbol : TypeSymbol, ITupleTypeSymbol
-        private ImmutableArray<TupleFieldSymbol> GetTupleFields(
-            ImmutableArray<string> elementNames, 
-            NamedTypeSymbol _underlyingType)
+        internal NamedTypeSymbol UnderlyingTupleType
```
Should probably add property IsTuple similar to IsAnonymousType.

VSadov replied:
I do not expect, yet, that tuple need a lot of special casing to add a helper. Can add later.

gafter commented on src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:
```diff
@@ -639,6 +626,63 @@ private BoundExpression BindExpressionInternal(ExpressionSyntax node, Diagnostic
+                var argumentSyntax = arguments[i];
+                var name = argumentSyntax.NameColon?.Name.Identifier.ValueText;
+
+                // names would typically all be there or none at all
```
Please add a PROTOTYPE comment, something like
```cs
// PROTOTYPE: we will need to check that all have names or none have names
```

AlekseyTs commented:
I don't like the idea of adding a new SymbolKind and TypeKind and have new symbol class inherit from NamedTypeSymbol. We have this with ErrorTypeSymbol, but this is an exception and we should try keeping it this way. Basically, symbols of distinct kind should not inherit from leaf classes, NamedTypeSymbol is a leaf class (ErrorTypeSymbol is an exception from the rule). If we want to have distinct Kind, the new class can inherit from TypeSymbol. On the other hand, we didn't introduce a new kind for Anonymous Types, perhaps we shouldn't do this for tuples too.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/TupleTypeSymbol.cs:
```diff
@@ -227,6 +225,13 @@ public override TResult Accept<TResult>(CSharpSymbolVisitor<TResult> visitor)
 
         internal override bool Equals(TypeSymbol t2, bool ignoreCustomModifiers, bool ignoreDynamic)
```
I think ContainingSymbol should return ContainingSymbol of the underlying tuple type.

VSadov replied:
agreed. does now.

gafter commented:
Please add a spec in `docs/features/tuples.md` and a working list of things not yet done in `docs/features/tuples.work.md`.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/TupleTypeSymbol.cs:
```diff
@@ -303,6 +308,158 @@ public override bool IsSealed
             }
         }
 
+        public override int Arity
```
DeclaredAccessibility should probably be the same as for the underlying type.

VSadov replied:
it is now

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/TupleTypeSymbol.cs:
```diff
@@ -303,6 +308,158 @@ public override bool IsSealed
+        {
+            get
+            {
+                return _underlyingType.IsComImport;
```
False

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/TupleTypeSymbol.cs:
```diff
@@ -303,6 +308,158 @@ public override bool IsSealed
+        {
+            get
+            {
+                return _underlyingType.IsWindowsRuntimeImport;
```
False?

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/TupleTypeSymbol.cs:
```diff
@@ -303,6 +308,158 @@ public override bool IsSealed
+        {
+            get
+            {
+                return _underlyingType.ShouldAddWinRTMembers;
```
False

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/TupleTypeSymbol.cs:
```diff
@@ -303,6 +308,158 @@ public override bool IsSealed
+        {
+            get
+            {
+                return _underlyingType.IsSerializable;
```
False?

VSadov replied:
underlying type in current implementation is actually serializable. I think this bit for the most part affects the run time and tuple will not exist at run time. It still feels like deferring to underlying is a better option.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/TupleTypeSymbol.cs:
```diff
@@ -303,6 +308,158 @@ public override bool IsSealed
+        {
+            get
+            {
+                return _underlyingType.IsInterface;
```
I am not sure it is appropriate to delegate to the _underlyingType for the 4 properties above.

VSadov replied:
i guess this is never an interface either

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/TupleTypeSymbol.cs:
```diff
@@ -303,6 +308,158 @@ public override bool IsSealed
+
         #region Use-Site Diagnostics
 
         internal override DiagnosticInfo GetUseSiteDiagnostic()
```
Why not simply delegate to the _underlyingType? This comment is for implementation.

VSadov replied:
fixed in next change

gafter commented on src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml:
```diff
@@ -1140,6 +1140,12 @@
   
+    <Node Name="BoundTupleCreationExpression" Base="BoundExpression">
+    <!-- Non-null type is required for this node kind -->
+    <Field Name="Type" Type="TypeSymbol" Override="true" Null="disallow"/>
```
I think that a tuple creation expression doesn't have a type. Or rather, there is probably a "raw" version without a type and a "cooked" version with a type where each element has been converted.

VSadov replied:
I think that would be a separate UnboundTupleCreationExpression if we do that

gafter commented on src/Compilers/CSharp/Portable/Emitter/Model/PEModuleBuilder.cs:
```diff
@@ -914,6 +914,9 @@ internal static Cci.IGenericParameterReference Translate(TypeParameterSymbol par
                     return Translate((TypeParameterSymbol)typeSymbol);
+
+                case SymbolKind.TupleType:
+                    return Translate(((TupleTypeSymbol)typeSymbol).UnderlyingTupleType, syntaxNodeOpt, diagnostics);
```
`// PROTOTYPE: need to write metadata for cross-assembly visibility of tuple member names`

gafter commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTupleTest.cs:
```diff
@@ -0,0 +1,313 @@
+  IL_0003:  ldc.i4.2
+  IL_0004:  ldc.i4.3
+  IL_0005:  newobj     ""System.Runtime.CompilerServices.ValueTuple<int, int>..ctor(int, int)""
+  IL_000a:  call       ""System.Runtime.CompilerServices.ValueTuple<int, >..ctor(int, )""
```
This doesn't look right. What is that thing after the comma?

VSadov replied:
this seems to be visualization issue. IL is correct.

VSadov commented:
test prtest/win/dbg/eta please

VSadov commented:
test this please

VSadov commented:
test prtest/win/dbg/eta please
Open to Internal Commit Pump not running

VSadov commented:
I think I have addressed most concerns and I am going to merge this to the feature branch since it becomes blocking.
I am however open to more suggestions. There is definitely more to be done.

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:
```diff
@@ -639,6 +626,89 @@ private BoundExpression BindExpressionInternal(ExpressionSyntax node, Diagnostic
+                // this should be a parse error already.
+                var args = arguments.Count == 1 ?
+                    new BoundExpression[] { BindValue(arguments[0].Expression, diagnostics, BindValueKind.RValue) } :
+                    new BoundExpression[0];
```
I think we have singletons for empty arrays in SpecializedCollections

VSadov replied:
fixed that in new change already.

AlekseyTs commented on src/Compilers/CSharp/Portable/Emitter/Model/PEModuleBuilder.cs:
```diff
@@ -1090,6 +1091,10 @@ internal override Cci.IMethodReference Translate(MethodSymbol symbol, Diagnostic
             }
+            else if (container.IsTupleType)
+            {
+                container = ((TupleTypeSymbol)container).UnderlyingTupleType;
```
I don't think this does anything useful. We actually want to fish-out the underlying method symbol.

AlekseyTs commented on src/Compilers/CSharp/Portable/Emitter/Model/PEModuleBuilder.cs:
```diff
@@ -1087,6 +1091,10 @@ internal override Cci.IMethodReference Translate(MethodSymbol symbol, Diagnostic
             }
+            else if (container.IsTupleType)
+            {
+                container = ((TupleTypeSymbol)container).UnderlyingTupleType;
```
I don't think this does anything useful. We actually want to fish-out the underlying method symbol.

AlekseyTs replied:
Do we need to intercept translation of fields? If we don't expect to run into them, we should still check and probably throw.

VSadov replied:
I thought about that, but decided not to touch that for now. We currently rewrite fields during lowering, but I think we might have to stop doing that. Debugging and EnC scenarios might require that.

VSadov replied:
And yes, intercepting methods is already removed in the other change. Tuples do not own methods.

VSadov replied:
I will add an assert for unexpected tuple fields in translator.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/TupleTypeSymbol.cs:
```diff
@@ -136,6 +136,14 @@ internal sealed override bool IsManagedType
             }
         }
 
+        public override bool IsTupleType
```
sealed?

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/TupleTypeSymbol.cs:
```diff
@@ -138,6 +136,14 @@ internal sealed override bool IsManagedType
             }
         }
 
+        public override bool IsTupleType
```
Sealed?

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/TupleTypeSymbol.cs:
```diff
@@ -174,7 +180,7 @@ public override SymbolKind Kind
             get
             {
-                return SymbolKind.TupleType;
+                return _underlyingType.Kind;
```
This should be SymbolKind.NamedType. Underlying could be an error type, but this symbol doesn't inherit from ErrorTypeSymbol. So, SymbolKind.NamedType is the only valid kind for it.

VSadov replied:
will change.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/TupleTypeSymbol.cs:
```diff
@@ -182,7 +188,7 @@ public override TypeKind TypeKind
             get
             {
-                return TypeKind.Tuple;
+                return _underlyingType.TypeKind;
```
This should probably be adjusted to not return TypeKind.Error

VSadov replied:
I think it can only be struct or class. Requirement to have fields rules out enums, delegates and the like.
Will change to be a struct if underlying is a struct, class otherwise.

VSadov replied:
Actually it may be better to dafault to a struct in broken scenarios since most common correct case will be struct.
